### PR TITLE
Better equality testing

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose

This PR improves #82 by using `numpy.can_cast` to test whether we can use `numpy.testing.assert_allclose` or we should test for equality instead.
This more general approach now allows for other JSON-serializable objects to be stored and compared in the reference files. A good example would be a single string value.

## Expected time until merged

2-3 days.

## Type of change

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing

I added a test.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

File: baseclasses/testing/pyRegTest.py [hunk 1]
```diff
@@ -263,23 +263,25 @@ def par_add_norm(self, name, values, **kwargs):
...
             full_name = name
         msg = f"Failed value for: {full_name}"
-        # special case if we're comparing list of strings
-        if isinstance(reference, list) and isinstance(reference[0], str):
-            self.assert_equal(actual, reference, name)
-        else:
-            numpy.testing.assert_allclose(actual, reference, rtol=rtol, atol=atol, err_msg=msg)
+        numpy.testing.assert_allclose(actual, reference, rtol=rtol, atol=atol, err_msg=msg)
 
-    def assert_equal(self, actual, reference, name):
```
Comment (marcomangano) [inline baseclasses/testing/pyRegTest.py:275]: What is the difference between `name` and `full_name` here?

File: baseclasses/testing/pyRegTest.py [hunk 2]
```diff
@@ -263,23 +263,25 @@ def par_add_norm(self, name, values, **kwargs):
...
+        """This is basically a wrapper on numpy.testing.assert_allclose with a generated error message"""
         if full_name is None:
             full_name = name
         msg = f"Failed value for: {full_name}"
-        # special case if we're comparing list of strings
-        if isinstance(reference, list) and isinstance(reference[0], str):
-            self.assert_equal(actual, reference, name)
-        else:
-            numpy.testing.assert_allclose(actual, reference, rtol=rtol, atol=atol, err_msg=msg)
+        numpy.testing.assert_allclose(actual, reference, rtol=rtol, atol=atol, err_msg=msg)
```
Comment (marcomangano) [inline baseclasses/testing/pyRegTest.py:270]: So now is on the user to call the right function, instead of figuring it out with an if check, right?

**Review by sseraj — COMMENTED** [review]
Looks good except for one typo

File: baseclasses/testing/pyRegTest.py [hunk 3]
```diff
@@ -263,23 +263,25 @@ def par_add_norm(self, name, values, **kwargs):
...
+        if full_name is None:
+            full_name = name
+        # special case where we treat tuples and lists the same
+        # we compare each element
         if isinstance(reference, (list, tuple)):
             for i, j in zip(actual, reference):
                 # cast both to tuple
                 if i != j:
-                    raise AssertionError(f"{i}, {j}")
+                    raise AssertionError(f"The elements do not match! Expected {j}, but got {j} instead.")
```
[the remark this answers is not in the document]
Comment (nwu63) [reply baseclasses/testing/pyRegTest.py]: Fixed

File: baseclasses/testing/pyRegTest.py (hunk 1 above)
Comment (nwu63) [reply baseclasses/testing/pyRegTest.py:275]: `name` is the name of this particular key, while `full_name` includes all the keys in a nested dictionary. Say for example you have
```
{"a": {"b": 1}}
```
`name` would be `b` but `full_name` would be `a: b`.

To be honest I think this stuff is only partially working, we should probably fix it at some point.

File: baseclasses/testing/pyRegTest.py (hunk 2 above)
Comment (nwu63) [reply baseclasses/testing/pyRegTest.py:270]: No, this is still detected automatically, but it's done in `_add_values` instead. I also switched from checking dtype to using numpy which is more reliable.

**Review by sseraj — APPROVED** [review]

**Pull request merged**